File: LongestSubstringwithAtLeastKRepeatingCharacters.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution {
public:
    int longestSubstring(string s, int k) {
        
        int i = 0, result = 0;
        while(i + k <= s.size()) {
        	
        	vector<int> m(26, 0);
        	int mask = 0, max_index = i;
			for(int j = i; j < s.size(); j++) {
				int t = s[j] - 'a';
				m[t]++;
				if(m[t] < k) mask |= 1 << t;
				else mask &= ~(1 << t);
				
				if(mask == 0) {
					result = max(result, j - i + 1);
					max_index = j;
				} 
			}
			
			i = max_index + 1;
		}
		
		return result;
        
    }
};

int main() {
	
	string s;
	cin>>s;
	
	int k;
	cin>>k;
	
	Solution *solution = new Solution();
	cout<<solution->longestSubstring(s, k)<<endl;
	
	return 0;
}

File: LongestSubstringwithAtLeastKRepeatingCharacters.py
class Solution:
    def longestSubstring(self, s, k):
        """
        :type s: str
        :type k: int
        :rtype: int
        """
        res = 0
        for h in range(1, 27):
            i, j = 0, 0
            unique = 0
            noLessThenK = 0
            m = {}
            while j < len(s):
                if unique <= h:
                    if s[j] not in m:
                        m[s[j]] = 0
                        unique += 1
                    m[s[j]] += 1
                    if m[s[j]] == k:
                        noLessThenK += 1
                    j += 1
                else:
                    if m[s[i]] == k:
                        noLessThenK -= 1
                    m[s[i]] -= 1
                    if m[s[i]] == 0:
                        unique -= 1
                        m.pop(s[i])
                    i += 1
                if unique == h and noLessThenK == h:
                    res = max(res, j - i)
        return res


if __name__ == '__main__':
    s = input()
    k = int(input())
    solution = Solution()
    print(solution.longestSubstring(s, k))
